pkg/agent/controller/bgp/controller.go

```diff
@@ -0,0 +1,954 @@
+		if err := c.watchSecretChanges(stopCh); err != nil {
+			klog.ErrorS(err, "Watch Secret error", "secret", types.BGPPolicySecretName)
+		}
+	}, time.Second*10, stopCh)
+
+	go wait.Until(c.worker, time.Second, stopCh)
+
+	<-stopCh
+}
+
+func (c *Controller) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *Controller) processNextWorkItem() bool {
+	defer c.queue.Done(key)
+
+	if err := c.syncBGPPolicy(); err == nil {
+		// If no error occurs we Forget this item, so it does not get queued again until another change happens.
+		c.queue.Forget(key)
+	} else {
+		// Put the item back on the work queue to handle any transient errors.
+		c.queue.AddRateLimited(key)
+		klog.ErrorS(err, "Syncing BGPPolicy failed, requeue")
+	}
+	return true
+}
+
+// filterBGPPolicies filters the BGPPolicies applied to the current Node.
+func (c *Controller) filterBGPPolicies() (map[string]*v1alpha1.BGPPolicy, error) {
+	allBPs, err := c.bgpPolicyLister.List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+
+	bpMap := make(map[string]*v1alpha1.BGPPolicy)
+	for _, bp := range allBPs {
+		if c.matchedCurrentNode(bp) {
+			bpMap[bp.GetName()] = bp
+		}
+	}
+	return bpMap, nil
+}
+
+func (c *Controller) syncBGPPolicy() error {
+	startTime := time.Now()
+	defer func() {
+		klog.V(2).InfoS("Finished syncing BGPPolicy", "durationTime", time.Since(startTime))
+	}()
+
+	// Get all available BGPPolicies applied to the current Node.
+	allBGPPolicies, err := c.filterBGPPolicies()
+	if err != nil {
+		return err
+	}
+
+	allBGPPolicyNames := sets.KeySet(allBGPPolicies)
+	preState := c.bgpPolicyState
+	var curState *bgpPolicyState
+	var effectiveBGPPolicyName string
+	var needUpdateBGPServer bool
+
+	if preState == nil {
+		if allBGPPolicyNames.Len() == 0 {
+			return nil
+		}
+		// If there is no effective BGPPolicy in the last sync, select a random available BGPPolicy as the effective one.
+		effectiveBGPPolicyName, _ = allBGPPolicyNames.PopAny()
+		curState = &bgpPolicyState{bgpPolicy: effectiveBGPPolicyName}
+		needUpdateBGPServer = true
+	} else {
+		// If there is an effective BGPPolicy in the last sync, check if it is still included in the available BGPPolicies.
+		if !allBGPPolicyNames.Has(preState.bgpPolicy) {
+			// If there is no available BGPPolicy, stop the previous BGP server, clean up the state and return.
+			if allBGPPolicyNames.Len() == 0 {
+				if err := preState.bgpServer.Stop(c.ctx); err != nil {
+					return err
+				}
+				c.bgpPolicyState = nil
+				return nil
+			}
+			// Select a new effective BGPPolicy from the available ones.
+			effectiveBGPPolicyName, _ = allBGPPolicyNames.PopAny()
+			curState = &bgpPolicyState{bgpPolicy: effectiveBGPPolicyName}
+			needUpdateBGPServer = true
+		} else {
+			// Retain the effective BGPPolicy in the last sync.
+			effectiveBGPPolicyName = preState.bgpPolicy
+			curState = &bgpPolicyState{bgpPolicy: effectiveBGPPolicyName}
+		}
+	}
+	bgpPolicy := allBGPPolicies[effectiveBGPPolicyName]
```

Comment:
I think I understand what it acheives with this code. However, this still doesn't ensure the same policy is enforced after a restart.
A simpler and stable solution could be just sorting policies in `filterBGPPolicies` and return the oldest policy, make name it `getBGPPolicy` then.

Reply:
Done